node/src/exchange.rs

```diff
@@ -31,21 +34,99 @@ pub fn new_behaviour(network: &str) -> Behaviour {
 #[derive(Clone, Copy, Debug, Default)]
 pub struct HeaderCodec;
 
+impl HeaderCodec {
+    async fn read_raw_message<T, R>(
+        io: &mut T,
+        buf: &mut Vec<u8>,
+        max_len: usize,
+    ) -> io::Result<Option<R>>
+    where
+        T: AsyncRead + Unpin + Send,
+        R: Message + Default,
+    {
+        let mut read = buf.len(); // buf might have data from previous iterations
+        let len = loop {
+            if let Ok(len) = prost::decode_length_delimiter(&buf[..read]) {
+                break len;
+            }
+
+            if read >= PROTOBUF_MAX_LENGTH_DELIMITER_LEN {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    ReadHeaderError::VarintOverflow,
+                ));
+            }
+
+            buf.resize(PROTOBUF_MAX_LENGTH_DELIMITER_LEN, 0);
```

Comment:
this should be done only if `buf.len()` is less than `PROTOBUF_MAX_LENGTH_DELIMITER_LEN`, otherwise it will discard data.

```suggestion
            if buf.len() < PROTOBUF_MAX_LENGTH_DELIMITER_LEN {
                buf.resize(PROTOBUF_MAX_LENGTH_DELIMITER_LEN, 0);
            }
```

Reply:
this can be removed now because of the buf resizing before the loop

Reply:
For previous implementation I believe it held that we had at most `PROTOBUF_MAX_LENGTH_DELIMITER_LEN` bytes cached, since after varint read (to 10byte buffer), we either had complete message and passed the remainder to next iteration; or we didn't and we read exact number of bytes needed, giving us empty buffer on next invocation.
With resizing to 512 before the loop whole thing changes though, so removing.